environs/manual/winrmprovisioner/winrmprovisioner.go

```diff
@@ -0,0 +1,581 @@
+		Secure:   false,
+	}); err != nil {
+		return errors.Annotatef(err, "cannot create a new http winrm client ")
+	}
+
+	logger.Infof("Trying http client as user %s on %s", args.Host, args.User)
+	if err = args.WClient.Ping(); err != nil {
+		logger.Debugf("WinRM unsecure listener is not enabled on %s", args.Host)
+		return errors.Annotatef(err, "cannot provision, because all winrm default connections failed")
+	}
+
+	defClient := args.WClient //save http client here in case we need to revert back
+	logger.Infof("Trying to enable https client certificate authenticaion")
+	if args.WClient, err = enableCertAuth(args); err != nil {
+		logger.Infof("Cannot enable client auth cert authentication for winrm")
+		logger.Infof("Reverting back to usecure client interaction")
+		// everything failed expept the http clinet
+		// so use this to provision the machine
+		args.WClient = defClient
+		return nil
+	}
+	logger.Infof("Client certs are installed and setup on the %s with user %s", args.Host, args.User)
+
+	if err = args.WClient.Ping(); err == nil {
+		logger.Infof("Winrm client authentication is now enabled on the host %s with user %s", args.Host, args.User)
+		return nil
+	}
+
+	logger.Infof("Winrm https connection is broken, can't retrive a response")
+	logger.Infof("Reverting back to usecure client interactions")
+	// everything failed expept the http clinet
+	// so use this to provision the machine
+	args.WClient = defClient
+
+	return nil
+
+}
+
+// enableCertAuth enables https cert auth interactions
+// with the winrm listener and returns the client
+func enableCertAuth(args *manual.ProvisionMachineArgs) (manual.WinrmClientAPI, error) {
+	var stderr bytes.Buffer
+	pass := args.WClient.Password()
+
+	scripts, err := bindInitScripts(pass, args.WKeys)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	for _, script := range scripts {
+		err = args.WClient.Run(script, args.Stdout, &stderr)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		if stderr.Len() > 0 {
+			return nil, errors.Errorf(
+				"encountered error executing cert auth script: %s",
+				stderr.String(),
+			)
+		}
+	}
+
+	cfg := winrm.ClientConfig{
+		User:    args.User,
+		Host:    args.Host,
+		Key:     args.WKeys.ClientKey(),
+		Cert:    args.WKeys.ClientCert(),
+		Timeout: 25 * time.Second,
+		Secure:  true,
+	}
+
+	caCert := args.WKeys.CACert()
+	if caCert == nil {
+		logger.Infof("Skipping winrm CA validation")
+		cfg.Insecure = true
+	} else {
+		cfg.CACert = caCert
+	}
+
+	return winrm.NewClient(cfg)
+}
+
+// bindInitScripts creates a series of scripts in a standard
+// (utf-16-le, base64)format for passing to the winrm conn to be executed remotely
+// we are doing this instead of one big script because winrm supports
+// just 8192 length commands. We know we have an amount of prefixed scripts
+// that we want to bind for the init process so create an array of scripts
+func bindInitScripts(pass string, keys *winrm.X509) ([3]string, error) {
```

Comment:
This code isn't on any hot path, so how the memory is allocated is not important. Maintainability is more important in general. Please use a slice; the caller should not care how many elements there are. As such, exposing it is creating too tight a dependency from callers to the implementation details.

(Also, FWIW: you can preallocate a slice and assign by index just as you do below. The compiler might even be smart enough to elide bounds checking, but even if it's not, it's going to have no noticeable impact on the runtime of the program.)